GuilhermeMota93/woocommerce-android
# Feat/products 2707 simple products details add text

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION
This PR contains the following:
- refactor of **"add new product"** in progress details
- initial cards for **title** and **description**

Tackling the next item on my list, the cards for title and description, I felt the need of refactoring (a tiny bit) some of my existing work. Up until here, I was handling the different state updates individually, which was not great in terms of code cleanse and general simplicity. 

I followed the existing idea of a `product draft` and adapted to the new flow of **add new product**.  As I got more and more familiar with the codebase, this made sense as a next step.

This approach would help achieve 2 things:
- keep parity with the existing details flow - other contributors can pick up on this new work and understand it easier as it matches the existing "pattern" of a product draft
- keep the **"add new product"** updated details in one place, making it easier to update and maintain 👍 

I called this new piece of refactored work the (very similar to the already existing `Product`) `AddNewProduct` with a parameter inside the `viewState` called `addNewProductDraft`. This `addNewProductDraft` is updated via the already known `viewState.copy()` call, which triggers the correct observers to then update the UI accordingly.

So far, so good 🎉 

Now for the details cards!

Also took some inspiration from the existing logic for the **edit product cards** and created a new builder called `AddProductDetailCardBuilder`. This is a simple builder accounting only for the **title** and **description** steps.

This new builder is called by the `updateAddNewProductCards()`, that gets triggered on each view state update - this way it always keeps the cards up to date.

Let me know your thoughts on it 🙏 

**NOTE**
There is a small caveat in this work.

My approach was to reuse the existing card UI from the details page. From the [figma resources](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=146%3A11&viewport=-2591%2C1628%2C0.6194302439689636) it shows the `Description` block being a bit different than the existing UI. 

I will reach out to @Garance Rey to double check what is the desired UI for this component in the **Add Product** flow 👍 

## TESTING
- Click on the PRODUCT bottom nav bar button
    - Wait for the screen to finish loading
    - Click on the FAB button
    - Notice the product detail for a new product flow, with title and description cards
    - Click on the title placeholder
    - Notice the capability of editing
    - Click on the description placeholder
    - Notice that it opens the `AztecEditorFragment` for description input with validations
    - Pressing back notice that the state being presented correctly, with the initial changes
    - Click on the add image container
    - After adding an image, navigate back to the Product Detail screen
    - Notice the state of title and description being presented correctly, with the initial changes

## SCREENSHOTS

<image src="https://user-images.githubusercontent.com/7058393/90832881-b4432c00-e33e-11ea-8e84-4f3de6c24ce8.png" width="320"/>

<image src="https://user-images.githubusercontent.com/7058393/90832909-c624cf00-e33e-11ea-8abc-5e41fbe9f6d5.gif" width="320"/> <image src="https://user-images.githubusercontent.com/7058393/90832924-cc1ab000-e33e-11ea-8269-6e6ea9205e69.gif" width="320"/>
